sbap/docking.py

```diff
@@ -0,0 +1,92 @@
+import pathlib
+import subprocess
+import tempfile
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import rdkit.Chem
+
+
+@dataclass
+class DockingConfig:
+    center_x: float
+    center_y: float
+    center_z: float
+    size_x: float
+    size_y: float
+    size_z: float
+    exhaustiveness: int
+
+
+class Dockerizer(ABC):
+    @abstractmethod
+    def dock(
+            self,
+            protein: rdkit.Chem.Mol,
+            ligands: list[rdkit.Chem.Mol],
+            docking_config: DockingConfig,
+    ) -> list[rdkit.Chem.Mol]:
+        pass
+
+
+class SminaDockerizer(Dockerizer):
+    def dock(
+            self,
+            protein: rdkit.Chem.Mol,
+            ligands: list[rdkit.Chem.Mol],
+            docking_config: DockingConfig,
+    ) -> list[rdkit.Chem.Mol]:
+        with tempfile.TemporaryDirectory() as directory_str:
+            directory = pathlib.Path(directory_str)
+            protein_path = directory.joinpath("protein.pdb")
+            rdkit.Chem.MolToPDBFile(protein, str(protein_path))
+            return [self._dock_single_ligand(ligand, directory, docking_config, protein_path) for ligand in ligands]
+
+    def _dock_single_ligand(
+            self,
+            ligand: rdkit.Chem.Mol,
+            directory: pathlib.Path,
+            docking_config: DockingConfig,
+            protein_path: pathlib.Path,
+    ) -> rdkit.Chem.Mol:
+        ligand_path = directory.joinpath("ligand.mol")
+        mol2_ligand_path = directory.joinpath("ligand.mol2")
+        docked_molecule_path = directory.joinpath("docked.mol2")
+        rdkit.Chem.MolToMolFile(ligand, str(ligand_path))
+        self._run_obabel(ligand_path, mol2_ligand_path, directory)
```

Comment:
wydaje mi się, że tu wcześniej chcielibyśmy jeszcze wywołać ten kawałek kodu, który był w notebooku z zajęć:
def optimize_conformation(mol):
    mol = Chem.AddHs(mol)  # Adds hydrogens to make optimization more accurate
    AllChem.EmbedMolecule(mol)  # Adds 3D positions
    AllChem.MMFFOptimizeMolecule(mol)  # Improves the 3D positions using a force-field method
    return mol
bo myśmy dostali te związki w 2D, a chcemy w 3D je mieć

Reply:
Jasne, zapomniałem o tym.